core/src/main/java/org/testcontainers/utility/MountableFile.java

```diff
@@ -0,0 +1,198 @@
+            resolvedPath = extractClassPathResourceToTempLocation(this.path);
+        } else {
+            resolvedPath = unencodeResourceURIToFilePath(path);
+        }
+
+        if (SystemUtils.IS_OS_WINDOWS) {
+            resolvedPath = PathUtils.createMinGWPath(resolvedPath);
+        }
+
+        return resolvedPath;
+    }
+
+    @NotNull
+    private static URL getClasspathResource(@NotNull final String resourcePath, @NotNull final Set<ClassLoader> classLoaders) {
+
+        // try context and system classloaders as well
+        classLoaders.add(Thread.currentThread().getContextClassLoader());
+        classLoaders.add(ClassLoader.getSystemClassLoader());
+        classLoaders.add(MountableFile.class.getClassLoader());
+
+        for (final ClassLoader classLoader : classLoaders) {
+            URL resource = classLoader.getResource(resourcePath);
+            if (resource != null) {
+                return resource;
+            }
+
+            // Be lenient if an absolute path was given
+            if (resourcePath.startsWith("/")) {
+                resource = classLoader.getResource(resourcePath.replaceFirst("/", ""));
+                if (resource != null) {
+                    return resource;
+                }
+            }
+        }
+
+        throw new IllegalArgumentException("Resource with path " + resourcePath + " could not be found on any of these classloaders: " + classLoaders);
+    }
+
+    private static String unencodeResourceURIToFilePath(@NotNull final String resource) {
+        try {
+            // Convert any url-encoded characters (e.g. spaces) back into unencoded form
+            return new URI(resource).getPath();
+        } catch (URISyntaxException e) {
+            throw new IllegalStateException(e);
+        }
+    }
+
+    /**
+     * Extract a file or directory tree from a JAR file to a temporary location.
+     * This allows Docker to mount classpath resources as files.
+     *
+     * @param hostPath the path on the host, expected to be of the format 'file:/path/to/some.jar!/classpath/path/to/resource'
+     * @return the path of the temporary file/directory
+     */
+    private String extractClassPathResourceToTempLocation(final String hostPath) {
+        File tmpLocation = new File(".testcontainers-tmp-" + Base58.randomString(5));
+        //noinspection ResultOfMethodCallIgnored
+        tmpLocation.delete();
+
+        String jarPath = hostPath.replaceFirst("jar:", "").replaceFirst("file:", "").replaceAll("!.*", "");
+        String urldecodedJarPath;
+        try {
+            urldecodedJarPath = URLDecoder.decode(jarPath, Charsets.UTF_8.name());
+        } catch (UnsupportedEncodingException e) {
+            throw new IllegalArgumentException("Could not URLDecode path with UTF-8 encoding: " + hostPath, e);
+        }
+        String internalPath = hostPath.replaceAll("[^!]*!/", "");
+
+        try (JarFile jarFile = new JarFile(urldecodedJarPath)) {
```

Comment:
starting from Java 7 we can use Jar (aka zip) FileSystem provider:
http://docs.oracle.com/javase/7/docs/technotes/guides/io/fsp/zipfilesystemprovider.html

it also might contain path escaping OOTB (not sure here, have to re-check)